src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -225,19 +268,259 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+            DataServiceResponse response = new DataServiceResponse(headers, (int)this.batchResponseMessage.StatusCode, operationResponses, this.IsBatchRequest);
+
+            try
+            {
+                // We are setting the MaxProtocolVersion to v4.01 when reading deep insert responses.
+                this.RequestInfo.Context.MaxProtocolVersion = ODataProtocolVersion.V401;
+                Version responseVersion;
+                HandleResponse(
+                    this.RequestInfo,
+                    (HttpStatusCode)this.batchResponseMessage.StatusCode,
+                    this.batchResponseMessage.GetHeader(XmlConstants.HttpODataVersion),
+                    () => this.ResponseStream,
+                    throwOnFailure: true,
+                    out responseVersion);
+
+                HeaderCollection headers = new HeaderCollection(this.batchResponseMessage);
+
+                ResponseInfo responseInfo = this.CreateResponseInfo(entityDescriptor);
+                HttpWebResponseMessage responseMessageWrapper = new HttpWebResponseMessage(
+                    headers,
+                    this.batchResponseMessage.StatusCode,
+                    () => this.ResponseStream);
+
+                ODataMaterializerContext materializerContext = new ODataMaterializerContext(responseInfo, base.MaterializerCache);
+                entry = ODataReaderEntityMaterializer.ParseSingleEntityPayload(responseMessageWrapper, responseInfo, entityDescriptor.Entity.GetType(), materializerContext);
+
+                this.headers = headers;
+                this.topLevelEntry = entry;
+            }
+            catch (DataServiceRequestException)
+            {
+                throw;
+            }
+            catch (InvalidOperationException ex)
+            {
+                HeaderCollection headers = new HeaderCollection(this.batchResponseMessage);
+                int statusCode = this.batchResponseMessage == null ? (int)HttpStatusCode.InternalServerError : (int)this.batchResponseMessage.StatusCode;
+                DataServiceResponse dataServiceResponse = new DataServiceResponse(headers, statusCode, Enumerable.Empty<OperationResponse>(), this.IsBatchRequest);
+                throw new DataServiceRequestException(Strings.DataServiceException_GeneralError, ex, dataServiceResponse);
+            }
+
+            HandleDeepInsertResponseInternal(entry, true, entityDescriptor, null);
+
+            return response;
+        }
+
+        private void HandleDeepInsertResponseInternal(MaterializerEntry entry, bool isTopLevelDescriptor, Descriptor descriptor, OperationResponse parentOperationResponse)
+        {
+            OperationResponse response = this.CreateOperationResponse(entry, isTopLevelDescriptor, descriptor, parentOperationResponse);
+
+            List<Descriptor> relatedDescriptors = this.BulkUpdateGraph.GetRelatedDescriptors(descriptor);
+
+            foreach (IMaterializerState nestedItem in entry?.NestedItems)

```

Comment:
Given your explanation that `entry` can never be null, why retain the `?` operator? It's a distraction to reviewers and maintainers of the code...

Reply:
The top level `entry` can never be `null`.
I am not sure about the nested entries